# A Python Quick Start Tutorial
# Lesson \#1: Basic data types and functions
## by Peter Mackenzie-Helnwein
University of Washington, Seattle, WA

pmackenz@uw.edu          
https://www.ce.washington.edu/facultyfinder/peter-mackenzie-helnwein

## Resources

   1. Python Docs: https://docs.python.org/3/
   
   1. Python Tutorial (comprehensive): https://docs.python.org/3/tutorial/index.html
   
   1. Python Library Reference (the nitty-gritty details): https://docs.python.org/3/library/index.html
   
   1. Everything else: http://google.com
   

## Discussing questions from the self-study assignment

This section is to discuss your questions


### Exercise 1

**Theory**:
Stress transformation

$$
\begin{aligned}
\sigma_{x}' &= \sigma_{x} \cos^2\theta + \sigma_{y} \sin^2\theta + 2\tau_{xy} \sin\theta \cos\theta \\
\sigma_{y}' &= \sigma_{x} \sin^2\theta + \sigma_{y} \cos^2\theta - 2\tau_{xy} \sin\theta \cos\theta \\
\tau_{xy}'  &= (\sigma_{y} - \sigma_{x}) \sin\theta \cos\theta + \tau_{xy} (\cos^2\theta - \sin^2\theta) \\
\end{aligned}
$$

**Given**:
Stress state:

$$
\sigma_{x} = 12~ksi~,
~~~
\sigma_{y} = -5.5~ksi~,
~~~
\tau_{xy} = 3.5~ksi 
$$

**Find**:
Components of stress in a rotated coordinate system for $\theta=25^\circ$.

In [8]:
# theta needs to be in radians
# sin and cos are not part of phython

from math import sin, cos, pi, radians

In [11]:
radians(90)

1.5707963267948966

In [29]:
from math import sin, cos, pi, radians
sigma_x = 12
sigma_y = -5.5
tau_xy = 3.5

theta = radians(25.)
sx = sigma_x * cos(theta)**2 + sigma_y * sin(theta)**2 + 2*tau_xy * sin(theta) * cos(theta)
sy = sigma_x * sin(theta)**2 + sigma_y * cos(theta)**2 - 2*tau_xy * sin(theta) * cos(theta)
txy = (sigma_y - sigma_x) * cos(theta) * cos(theta) + tau_xy * (cos(theta)**2 - sin(theta)**2)
print("sigma_x' = {:8.4f} ksi".format(sigma_x))
print("sigma_y' = {:8.4f} ksi".format(sigma_y))
print("tau_xy' = {:8.4f} ksi".format(tau_xy))
print("sx' = {:8.4f} ksi".format(sx))
print("sy' = {:8.4f} ksi".format(sy))
print("txy' = {:8.4f} ksi".format(txy))

sigma_x' =  12.0000 ksi
sigma_y' =  -5.5000 ksi
tau_xy' =   3.5000 ksi
sx' =  11.5555 ksi
sy' =  -5.0555 ksi
txy' = -12.1246 ksi


### Exercise 2

Convert the calculation from exercise 1 to a function

In [34]:
def stressTransform(sigx, sigy, tauxy, theta):

    mytheta = radians(theta)

    sx  = sigx * cos(mytheta)**2 + sigy * sin(mytheta)**2 + 2*tauxy * sin(mytheta) * cos(mytheta)
    sy  = sigx * sin(mytheta)**2 + sigy * cos(mytheta)**2 - 2*tauxy * sin(mytheta) * cos(mytheta)
    txy = (sigy - sigx) * sin(mytheta) * cos(mytheta) + tauxy * (cos(mytheta)**2 - sin(mytheta)**2) 
    
    return (sx,sy,txy)

Test your function for 

* $\theta=0$ (what should the result look like?)
* $\theta=25^\circ$
* $\theta=180^\circ$

In [36]:
stressTransform(sigma_x, sigma_y, tau_xy, 0)

(12.0, -5.5, 3.5)

In [37]:
stressTransform(sigma_x, sigma_y, tau_xy, 25)

(11.555547135673642, -5.055547135673642, -4.453132243388171)

In [38]:
stressTransform(sigma_x, sigma_y, tau_xy, 180)

(12.0, -5.499999999999999, 3.500000000000002)

In [42]:
stress=stressTransform(sigma_x, sigma_y, tau_xy, 25)

### Exercise 3

Let's use this function to study stress transformation.  In order to do so,

1. Create a list of 81 evenly spaced $\theta$-values from $0^\circ$ to $180^\circ$.
2. Compute the transformed stress for each of these values and collect the results in three lists: 

       sigmax_values = ...
       sigmay_values = ...
       tauxy_values = ...

3. print a nice table with 4 columns: $\theta$, $\sigma_x'$, $\sigma_y'$, $\tau_{xy}'$

In [43]:
print(stress)

(11.555547135673642, -5.055547135673642, -4.453132243388171)


In [44]:
print("sigma_x = {:8.4f} ksi\nsigma_y = {:8.4f} ksi\ntau = {:8.4f} ksi".format(stress[0],stress[1],stress[2]))

sigma_x =  11.5555 ksi
sigma_y =  -5.0555 ksi
tau =  -4.4531 ksi


In [45]:
sigmax_values = []
sigmay_values = []
tauxy_values = []

# YOUR CODE HERE

In [48]:
theta = []

print("       theta       sigma_x       sigma_y        tau_xy ")
print("=======================================================")

for i in range(81):
    th = i*180/80;
    
    theta.append(th)
    (sx,sy,tau) = stressTransform(sigma_x, sigma_y, tau_xy, th)
    sigmax_values.append(sx)
    sigmay_values.append(sy)
    tauxy_values.append(tau)
    
    print("{:12.6f}, {:12.6f}, {:12.6f}, {:12.6f}".format(th,sx,sy,tau))

       theta       sigma_x       sigma_y        tau_xy 
    0.000000,    12.000000,    -5.500000,     3.500000
    2.250000,    12.247634,    -5.747634,     2.802694
    4.500000,    12.439794,    -5.939794,     2.088108
    6.750000,    12.575296,    -6.075296,     1.360648
    9.000000,    12.653304,    -6.153304,     0.624799
   11.250000,    12.673338,    -6.173338,    -0.114902
   13.500000,    12.635274,    -6.135274,    -0.853894
   15.750000,    12.539346,    -6.039346,    -1.587622
   18.000000,    12.386147,    -5.886147,    -2.311561
   20.250000,    12.176620,    -5.676620,    -3.021250
   22.500000,    11.912058,    -5.412058,    -3.712311
   24.750000,    11.594091,    -5.094091,    -4.380484
   27.000000,    11.224680,    -4.724680,    -5.021650
   29.250000,    10.806103,    -4.306103,    -5.631856
   31.500000,    10.340940,    -3.840940,    -6.207340
   33.750000,     9.832058,    -3.332058,    -6.744554
   36.000000,     9.282597,    -2.782597,    -7.240185
   38.250

### Exercise 4: Explore plotting in python

For this, check out https://matplotlib.org/3.1.1/gallery/lines_bars_and_markers/simple_plot.html#sphx-glr-gallery-lines-bars-and-markers-simple-plot-py

I copied the demo code below so you can modify it to plot $(\sigma_{x}'/\tau_{xy}')$ and  $(\sigma_{y}'/-\tau_{xy}')$ curves from the data collected in `sigmax_values`, `sigmay_values`, `tauxy_values`

In [ ]:
import matplotlib
import matplotlib.pyplot as plt
import numpy as np

# Data for plotting
t = np.arange(0.0, 2.0, 0.01)
s = 1 + np.sin(2 * np.pi * t)

fig, ax = plt.subplots()
ax.plot(t, s)

ax.set(xlabel='time (s)', ylabel='voltage (mV)',
       title='About as simple as it gets, folks')
ax.grid()

# I comment writing the figure to file ... makes this quicker
# fig.savefig("test.png")
plt.show()

**Remark**: This was just to wet your appetite! We will talk much more about `numpy` and `matplotlib` next week.

## Exercise 5: List versus dictionary

You could express the stress state as

1. three independent variables - the way we did it thus far

2. a list of three values: `stress = `$[\sigma_x, \sigma_y, \tau_{xy}]$ and replace

        sigma_x --> stress[0]
        sigma_y --> stress[2]
        tau_xy  --> stress[3]
        
3. a dictionary 

        stress = {'sigx': 12.0, 'sigy': -5.5, 'tau':3.50}
        
   and replace

        sigma_x --> stress['sigx']
        sigma_y --> stress['sigy']
        tau_xy  --> stress['tau']

**Discuss**:

1. Pros and cons for coding the equations
2. Pros and cons for coding the function(s)
3. How woud code and function change when switching from 2D to 3D (6 components instead of 3 components)

Feel free to implement an alternative form of stressTrannsform() that looks like this:

In [ ]:
def stressTransform(stress, theta):
    # YOUR CODE
    # stress2 is a list or dictionary, just as stress is
    
    return stress2

## Homework questions

We can all learn from your questions!